crates/core/tedge_api/src/messages.rs

```diff
@@ -473,65 +477,139 @@ impl From<SoftwareError> for Option<SoftwareModuleItem> {
-}
 
-#[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
-pub struct RestartOperationResponse {
-    pub id: String,
-    pub status: OperationStatus,
-}
+    /// Set the status of the command
+    pub fn with_status(mut self, status: CommandStatus) -> Self {
+        self.payload.status = status;
+        self
+    }
 
-impl<'a> Jsonify<'a> for RestartOperationResponse {}
+    /// Set the failure reason of the command
+    pub fn with_error(mut self, reason: String) -> Self {
+        self.payload.status = CommandStatus::Failed;
+        self.payload.reason = reason;
+        self
+    }
+
+    /// Return the MQTT topic identifier of the target
+    fn topic_id(&self) -> &EntityTopicId {
+        &self.target
+    }
 
-impl RestartOperationResponse {
-    pub fn new(req: &RestartOperationRequest) -> Self {
-        Self {
-            id: req.id.clone(),
-            status: OperationStatus::Executing,
+    /// Return the MQTT channel for this command
+    fn channel(&self) -> Channel {
+        Channel::Command {
+            operation: OperationType::Restart,
+            cmd_id: self.cmd_id.clone(),
         }
     }
 
-    pub fn with_status(self, status: OperationStatus) -> Self {
-        Self { status, ..self }
+    /// Return the MQTT topic for this command
+    fn topic(&self, schema: &MqttSchema) -> Topic {
+        schema.topic_for(self.topic_id(), &self.channel())
     }
 
-    pub fn topic() -> Topic {
-        Topic::new_unchecked(DEVICE_RESTART_RESPONSE_TOPIC)
+    /// Return the MQTT Message to register `restart` commands on a given target device
+    pub fn capability_message(schema: &MqttSchema, target: &EntityTopicId) -> Message {
+        let meta_topic = schema.topic_for(
+            target,
+            &Channel::CommandMetadata {
+                operation: OperationType::Restart,
+            },
+        );
+        let payload = "{}";
+        Message::new(&meta_topic, payload)
+            .with_retain()
+            .with_qos(QoS::AtLeastOnce)
     }
 
-    pub fn status(&self) -> OperationStatus {
-        self.status
+    /// Return the MQTT payload for this command
+    pub fn command_message(&self, schema: &MqttSchema) -> Message {
+        let topic = self.topic(schema);
+        let payload = self.payload.to_bytes();
+        Message::new(&topic, payload)
+            .with_qos(QoS::AtLeastOnce)
+            .with_retain()
+    }
+
+    /// Return the MQTT message to clear this command
+    pub fn clearing_message(&self, schema: &MqttSchema) -> Message {
+        let topic = self.topic(schema);
+        Message::new(&topic, vec![])
+            .with_qos(QoS::AtLeastOnce)
+            .with_retain()
+    }
+}
+
+/// Command to restart a device
+#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct RestartCommandPayload {
+    pub status: CommandStatus,
+
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub reason: String,
+}
```

Comment:
Is the `reason` field supposed to be not empty only when `status` is `CommandStatus::Failed`? If there are no other motivations for this, I suggest using [Serde's internally tagged enum representation](https://serde.rs/enum-representations.html#internally-tagged). Modified `CommandStatus` would look like this:

```rust
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Clone)]
#[serde(rename_all = "camelCase", tag = "status")]
pub enum CommandStatus {
    Init,
    Executing,
    Successful,
    Failed { reason: String },
}
```

And the `reason` field would automatically be serialised/deserialised and only be present when in `Failed` variant, which looks the same as the current structure.

One issue is that `CommandStatus` would no longer be copy, and also perhaps there would have to be more code changes.

`RestartCommandPayload would then simply be:

```suggestion
#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RestartCommandPayload {
    #[serde(flatten)]
    pub status: CommandStatus,
}
```